### DISCUSSION TITLE
Golang integration

### DISCUSSION THREAD

**Discussion by vit1251** [slint-ui/slint#503] [Q&A]
I have not found any mention of API or runtime for Golang.

Could you please clarify this topic: 

- Is this possible?
- Do you have any plan to do so in the near future?
- Workarounds?

**Comment by tronical** [comment]
I think supporting golang is possible. We don't have any plans to do so in the near future, but we're happy to engage in discussions evolving around the scope and how it could be implemented and incorporated into the existing architecture.

There are two general approaches that could be taken:

1. Provide an API/module to embed the run-time interpreter. That implies the ability to dynamically load .60 files, but it would be loosely typed on the golang API side (still strictly within .60 of course).
2. Provide a backend to the compiler to generate golang code. I suppose this would be integrated with `//go:generate`, or is there are an automatic mechanism available?

The two are not mutually exclusive, for C++ we also offer both. But I think (1) might make more sense to start with. The C++ API for the interpreter uses an internal C-API, but there might be some C++isms in there regarding constructors/destructors for some of the run-time data types such as brushes or images. Those might need working around.

What's your experience with golang and binding to other languages?

**Reply by ogoffart** [reply]
I think we should try to opt for (1) that generates go code. This is probably giving the best experience in term of type safety and auto-completion.

Now I am not an expert Go programmer, and we would probably need help to design the API.

How would the perfect API look like, to set and get properties, callbacks, models. Also need to consider communication with other threads/go routines and such.

**Comment by vit1251** [comment]
> What's your experience with golang and binding to other languages?

I've made several bindings to various languages like Python, Lua, Node and others. I think that Golang provide common FFI for function calls, but as far as I understand it should be a C calling agreement without any mangles and plusplusing.

**Comment by TristanAppDev** [comment]
Is there any news regarding Golang integration?  
Are there any plans as to when it might happen?

The combination of Slint's clean and simple dsl and Go's simplicity in general would be fantastic.
